app/src/tracing.h

```diff
@@ -0,0 +1,222 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "constants.h"
+
+#include <ccf/base_endpoint_registry.h>
+#include <ccf/endpoint_context.h>
+#include <functional>
+#include <regex>
+#include <sstream>
+#include <time.h>
+
+namespace scitt
+{
+  constexpr std::string_view REQUEST_ID_HEADER = "x-ms-request-id";
+  constexpr std::string_view CLIENT_REQUEST_ID_HEADER =
+    "x-ms-client-request-id";
+
+  const std::regex CLIENT_REQUEST_ID_REGEX("^[0-9a-zA-Z-]+");
+
+  thread_local std::string request_id;
+  thread_local std::optional<std::string> client_request_id;
+
+  struct AppData
+  {
+    std::string request_id;
+    std::optional<std::string> client_request_id;
+  };
+
+  void clear_trace_state()
+  {
+    request_id = "";
+    client_request_id = std::nullopt;
+  }
+
+  int diff_timespec_ms(
+    const struct timespec& time0, const struct timespec& time1)
+  {
+    return (time1.tv_sec - time0.tv_sec) * 1000 +
+      (time1.tv_nsec - time0.tv_nsec) / 1000000;
+  }
+
+  std::string create_request_id()
+  {
+    std::stringstream stream;
+    stream << std::hex << rand();
+    return stream.str();
+  }
```

Comment:
```suggestion
  std::string create_request_id()
  {
      return fmt::format("{:x}", rand());
  }
```

This might be a matter of personal preference, but I find this more readable.